pkg/storage/imagestore/imagestore.go

```diff
@@ -0,0 +1,2194 @@
+			}
+		}
+
+		// if the deleted blob is one with content
+		if dstRecord == blobPath {
+			// get next candidate
+			dstRecord, err := is.cache.GetBlob(digest)
+			if err != nil && !errors.Is(err, zerr.ErrCacheMiss) {
+				is.log.Error().Err(err).Str("blobPath", dstRecord).Msg("dedupe: unable to lookup blob record")
+
+				return err
+			}
+
+			// if we have a new candidate move the blob content to it
+			if dstRecord != "" {
+				/* check to see if we need to move the content from original blob to duplicate one
+				(in case of filesystem, this should not be needed */
+				binfo, err := is.storeDriver.Stat(dstRecord)
+				if err != nil {
+					is.log.Error().Err(err).Str("path", blobPath).Msg("rebuild dedupe: failed to stat blob")
+
+					return err
+				}
+
+				if binfo.Size() == 0 {
+					if err := is.storeDriver.Move(blobPath, dstRecord); err != nil {
+						is.log.Error().Err(err).Str("blobPath", blobPath).Msg("unable to remove blob path")
+
+						return err
+					}
+				}
+
+				return nil
+			}
+		}
+	}
+
+	if err := is.storeDriver.Delete(blobPath); err != nil {
+		is.log.Error().Err(err).Str("blobPath", blobPath).Msg("unable to remove blob path")
+
+		return err
+	}
+
+	return nil
+}
+
+type extendedManifest struct {
+	ispec.Manifest
+
+	Digest godigest.Digest
+}
+
+func (is *ImageStore) garbageCollect(repo string) error {
+	// gc untagged manifests and signatures
+	index, err := common.GetIndex(is, repo, is.log)
+	if err != nil {
+		return err
+	}
+
+	referencedByImageIndex := []string{}
+	cosignDescriptors := []ispec.Descriptor{}
+	artifactManifests := []extendedManifest{}
+
+	/* gather manifests references by multiarch images (to skip gc)
+	gather cosign and notation signatures descriptors */
+	for _, desc := range index.Manifests {
+		switch desc.MediaType {
+		case ispec.MediaTypeImageIndex:
+			indexImage, err := common.GetImageIndex(is, repo, desc.Digest, is.log)
+			if err != nil {
+				is.log.Error().Err(err).Str("repository", repo).Str("digest", desc.Digest.String()).
+					Msg("gc: failed to read multiarch(index) image")
+
+				return err
+			}
+
+			for _, indexDesc := range indexImage.Manifests {
+				referencedByImageIndex = append(referencedByImageIndex, indexDesc.Digest.String())
+			}
+		case ispec.MediaTypeImageManifest:
+			tag, ok := desc.Annotations[ispec.AnnotationRefName]
+			if ok {
+				// gather cosign references
+				if strings.HasPrefix(tag, "sha256-") && (strings.HasSuffix(tag, cosignSignatureTagSuffix) ||
+					strings.HasSuffix(tag, SBOMTagSuffix)) {
+					cosignDescriptors = append(cosignDescriptors, desc)
+
+					continue
+				}
+			}
+
+			manifestContent, err := common.GetImageManifest(is, repo, desc.Digest, is.log)
+			if err != nil {
+				is.log.Error().Err(err).Str("repo", repo).Str("digest", desc.Digest.String()).
+					Msg("gc: failed to read manifest image")
+
+				return err
+			}
+
+			if manifestContent.Subject != nil {
+				artifactManifests = append(artifactManifests, extendedManifest{
```

Comment:
```suggestion
				referrerManifests = append(referrerManifests, extendedManifest{
```
In case we still need this code after refactor.